### DISCUSSION TITLE
Speed BM25 when increase Top K retrieval result

### DISCUSSION THREAD

**Discussion by hieudx149** [castorini/pyserini#1228] [Q&A] [answered]
Hi,
I did some experiments, when i increase the number of results returned by bm25, the search speed slows down. I find this confusing because obviously as far as I know when doing a search bm25 is already doing a search on the entire corpus so returning 10, 20 or 10000 results has no effect on speed ?
Can you explain to me why the number of results returned affects the search speed ? Or did i miss something ?

**Comment by lintool** [comment, marked as answer]
Yes, it is well known that query latency increases as _k_ in top-_k_ retrieval increases. Here's a relatively recent survey that provides lots of details: https://www.nowpublishers.com/article/Details/INR-057

Insight is that top-_k_ docs are kept in the heap during query evaluation - the larger the _k_, the more "work" the algorithm needs to do. Modern algorithms are fast because they are able to (heuristically) answer this question: can this document possibly be in the top _k_? If no, I can skip it.

**Comment by hieudx149** [comment]
Hi @lintool ,
Thanks for your quick and detailed answer !!!
